[DOCS] data quality distribution analysis, working draft

Quantisan opened the pull request:
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

Quantisan committed:
create volume doc

draft

edit

update conclusion

add impact section

update examples and pitfalls

update conclusion

note TODOs and shorten data preview section

implement code samples

Quantisan committed:
check-in csv

Quantisan committed:
moved code snippets into CI

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Quantisan committed:
note TODO for Cloud screenshot

Quantisan committed:
Merge branch 'docs/data-quality-volume' of github.com:great-expectations/great_expectations into docs/data-quality-volume

* 'docs/data-quality-volume' of github.com:great-expectations/great_expectations:
  [pre-commit.ci] auto fixes from pre-commit.com hooks

Quantisan committed:
copy edit to shorten 'GX' references

Quantisan committed:
adding whitespace before tip

rachhouse committed:
remove link to future image so netlify builds, replace sup with small text

rachhouse committed:
Merge branch 'develop' into docs/data-quality-volume

Quantisan committed:
remove combo DQ example in Next Step

Quantisan committed:
remove ref to Github Actions blog, it's deprecated

Quantisan committed:
expand sample volume.csv to cover the course of a week

Quantisan committed:
separate Examples and Scenarios into 2 sections

Quantisan committed:
add volume_workflow.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Quantisan committed:
draft on DQ: distribution

Quantisan committed:
update data preview and key Expectations sections

Quantisan committed:
update pitfalls

rachhouse committed:
update code example and add python scripts to code under test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
reduce next steps section, add GX Cloud steps for example

rachhouse committed:
Merge branch 'develop' into docs/data-quality-volume

rachhouse committed:
add gif showing creation of volume Expectation

rachhouse committed:
add test for example workflow

rachhouse committed:
Merge branch 'develop' into docs/data-quality-volume

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
force gx imports first

Quantisan committed:
sample data use synthetic names

rachhouse committed:
fix tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Quantisan committed:
update Example code

rachhouse commented:
I'm not sure why we're not seeing the netlify preview now that the base is `develop`, I'm going to try merging in `develop` to see if it triggers the preview build.

rachhouse committed:
Merge branch 'develop' into docs/data-quality-distribution-analysis

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution.md:
````diff
@@ -1,8 +1,297 @@
+**Use Case**: Robustly monitors key statistics of the overall distribution, which is valuable for tracking metrics like median purchase amount or 90th percentile product ratings.
+
+```python
+gxe.expect_column_quantile_values_to_be_between(
````
Just a note for when you move these snippets into a tested script - Expectations should be in 1.0 CamelCaseSyntax.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution.md:
```diff
@@ -1,8 +1,297 @@
+
+In addition to the key distribution Expectations discussed above, GX offers a range of other Expectations that can be useful for validating and monitoring data distributions. Some examples include:
+
+- `ExpectColumnMedianToBeBetween`: Checks if the median value of a column is within a specified range. The median is the middle value when the data is sorted in ascending or descending order. This Expectation is useful for ensuring that the central tendency of the data falls within acceptable limits.
```
Is there a reason you opted not to provide examples and Expectation Gallery links for these Expectations? I think it would be helpful to also have example usage & links for these Expectations.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution.md:
```diff
@@ -1,8 +1,297 @@
+
+In this dataset, `purchase_amount` represents the amount spent by customers in various `product_category`. Analyzing the distribution of `purchase_amount` and other numerical columns can reveal insights into customer behavior and detect anomalies.
+
+## Key distribution Expectations
```
Currently, Expectations categorized as **Distribution** and **Numerical data** in the Expectations Gallery are relevant when discussing data distribution. I think Expectations from both groups can be profiled in this article, and that we can separate Expectations into some logical subgroups, like we did for schema:

(I've put their code names here for brevity, we can use the full Expectation name as the heading in the article).

Key distribution Expectations
* Column-level Expectations
  * ExpectColumnMaxToBeBetween
  * ExpectColumnMeanToBeBetween
  * ExpectColumnMedianToBeBetween
  * ExpectColumnMinToBeBetween
  * ExpectColumnQuantileValuesToBeBetween
  * ExpectColumnSumToBeBetween
  * ExpectColumnStdevToBeBetween
  * ExpectColumnKLDivergenceToBeLessThan
* Row-level Expectations
  * ExpectColumnValuesToBeBetween
  * ExpectColumnValueZScoresToBeLessThan

(I think `ExpectColumnPairValuesAToBeGreaterThanB` belongs in data integrity, since it's about the relationship of two columns).

rachhouse replied:
@Quantisan I see you resolved this comment, but I don't see all the Expectations now included in the section - was this just an oversight?

Quantisan replied:
Updated. I selected a few for a full showcase and kept the rest in the Additional section. Just so it's not feeling too repetitive for the reader.

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution.md:
```diff
@@ -1,8 +1,297 @@
+
+## Data preview
+
+The examples in this guide use a sample dataset of customer transactions, available as a [CSV file on GitHub](https://raw.githubusercontent.com/great-expectations/great_expectations/develop/tests/test_sets/learn_data_quality_use_cases/distribution_purchases.csv).
```
When you have this CSV added to the PR, I'll import it into our cloud Postgres instance for you as well.

rachhouse replied:
@Quantisan Thanks for adding the CSV! I'll get this imported into the cloud Postgres for you tomorrow.

rachhouse replied:
Alrighty, this data is now available from `postgresql+psycopg2://try_gx:try_gx@postgres.workshops.greatexpectations.io/gx_learn_data_quality` in the `distribution_purchases` table.

Quantisan committed:
remove ExpectColumnPairValuesAToBeGreaterThanB as it should belong to data integrity

https://github.com/great-expectations/great_expectations/pull/10440#discussion_r1783511289

Quantisan committed:
separate Expections to column-level and row-level

Quantisan committed:
tips for Expectations

Quantisan committed:
Expectations use 1.0 casing

Quantisan committed:
link to Expectations Gallery for the additionals

Quantisan committed:
move Expectation snippets into .py

Quantisan committed:
move workflow code into snippet

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Quantisan committed:
add distribution csv

Quantisan committed:
Merge branch 'docs/data-quality-distribution-analysis' of github.com:great-expectations/great_expectations into docs/data-quality-distribution-analysis

* 'docs/data-quality-distribution-analysis' of github.com:great-expectations/great_expectations:
  [pre-commit.ci] auto fixes from pre-commit.com hooks

Quantisan committed:
fix typo

Quantisan commented:
@rachhouse I could use your help here https://github.com/great-expectations/great_expectations/pull/10440/files#diff-2d7ecca302d265e03bd536578750b767ab748a63f5f4720edca1b9dc765563bcR93

rachhouse commented:
> @rachhouse I could use your help here https://github.com/great-expectations/great_expectations/pull/10440/files#diff-2d7ecca302d265e03bd536578750b767ab748a63f5f4720edca1b9dc765563bcR93

@Quantisan This was a nonobvious one - turns out the issue is that `reference_distribution` is currently defined inside the `suite.add_expectation()` call, and that's causing ruff to flag the syntax error. 

There's not a great way to define the reference distribution if we want to keep `add_expectation` LOC outside of the snippet. I think the best compromise is probably defining it directly in the Expectation definition:
```
suite.add_expectation(
    # <snippet name="docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution_resources/distribution_expectations.py model_data_drift">
    # Validate that KL divergence is below the specified threshold using the reference
    # distribution provided as partition_object.
    gxe.ExpectColumnKlDivergenceToBeLessThan(
        column="purchase_amount",
        partition_object={
        "bins": [0, 100, 500, 1000, 5000, 10000],
        "weights": [0.05, 0.25, 0.35, 0.25, 0.1]
    },
        threshold=0.1
    )
    # </snippet>
)
```

rachhouse commented on docs/docusaurus/docs/reference/learn/data_quality_use_cases/distribution_resources/distribution_expectations.py:
```diff
@@ -0,0 +1,146 @@
+)
+
+suite.add_expectation(
+    # TODO: Paul is not sure how to add multiple expectations to the suite at once, or some way to
```
I'm going to take a look at this tomorrow and see if there's a way to set this up so that the snippet doesn't have the extra character. We may have to assign the two Expectations to named variables and then add the variables using  `add_expectation(...)` outside of the snippet.

FYI, I checked in with Engineering on this topic, and currently, there's not a method that enables you to add multiple Expectations at once. I've generally been using a list comprehension or loop to do it.

Quantisan committed:
add 3 missign column-level expectations to Additional section

Quantisan committed:
move ExpectColumnMeanToBeBetween to a full showcase

Quantisan committed:
fix: use list comprehension for detecting_anomalies

Quantisan committed:
fix: ExpectColumnKlDivergenceToBeLessThan argument

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Quantisan commented:
Thanks for the review and help, @rachhouse. This is ready for another round of review.

Currently the CI is failing at `public-api-report` step and skipping the `docs-snippet` tests. I'm not sure what to do about `Items are missing from the public API: 5` https://github.com/great-expectations/great_expectations/actions/runs/11431557716/job/31800823075?pr=10440

rachhouse commented:
> Thanks for the review and help, @rachhouse. This is ready for another round of review.
> 
> Currently the CI is failing at `public-api-report` step and skipping the `docs-snippet` tests. I'm not sure what to do about `Items are missing from the public API: 5` https://github.com/great-expectations/great_expectations/actions/runs/11431557716/job/31800823075?pr=10440

Great, thanks! I'll take another look. I also haven't seen that public API error before, I'll dig into that for you.

rachhouse committed:
Merge branch 'develop' into docs/data-quality-distribution-analysis

rachhouse committed:
remove freshness placeholder

rachhouse committed:
reorder dq topics alphabetically

rachhouse committed:
check updates to code now under test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rachhouse committed:
add standard use case gif, fix public_api_report test by adding mentions of core Expectations to ITEMS_IGNORED_FROM_PUBLIC_API for now

rachhouse committed:
Merge branch 'develop' into docs/data-quality-distribution-analysis

rachhouse committed:
Merge branch 'docs/data-quality-distribution-analysis' of github.com:great-expectations/great_expectations into docs/data-quality-distribution-analysis

rachhouse committed:
fix image path

rachhouse committed:
K to the L

rachhouse committed:
Merge branch 'develop' into docs/data-quality-distribution-analysis

rachhouse reviewed (approved):
No changes came out of internal review (coordinated via Slack) - I'm approving this PR so that we can get it merged in.

rachhouse committed:
Merge branch 'develop' into docs/data-quality-distribution-analysis
